woocommerce/google-listings-and-ads
# Product Block Editor: Add the fundamental setup for the custom product blocks and map `Select` and `BooleanSelect` inputs to the custom select field block

### Changes proposed in this Pull Request:

To be compatible with Product Block Editor, this PR adds fundamental support:

- Adjust the `can_enqueue` of `BaseAsset` as it should allow `$enqueue_condition_callback` to be null.
- Adjust the `AttributesBlock` to register the custom blocks and the related assets.
- Adjust the frontend config to build the custom blocks.
- Add the structure of custom blocks.
- Update the `@woocommerce/dependency-extraction-webpack-plugin` package to get the support of the newer WC packages.

To show how to develop and use the custom product blocks, this PR:

- Implement the custom block of the product select field.
- Map `Select` and `BooleanSelect` inputs to the custom block of the select field.

### Screenshots:

#### 📷 Simple product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/3f38cc4f-9a2c-4916-82f9-3bf04e573ec0)

#### 📷 Variable product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/b080da60-eb51-4a1b-af6b-d7971cd25589)

#### 📷 Variation product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/5c8f7a03-9e5a-453b-930b-a431930b748c)

### Detailed test instructions:

#### 📌 Prepare test environment

1. Please refer to [the test preparation in PR 2151](https://github.com/woocommerce/google-listings-and-ads/pull/2151#prepare-test-environment).
1. `npm install`
1. `npm start`

#### 📌 Test the simple product type

1. Go to edit a simple product, e.g. the **Beanie with Logo** imported from WC's sample products
1. Switch to the **Organization** tab and find the **Google Listings & Ads** section.
1. Operate the selection fields and change their selected option.
1. Click on the **Update** button at the top-right of the page.
1. Refresh the page or switch back to the classical product editor to see if the changed options are saved.

#### 📌 Test the variable and variation product type

1. Go to edit a variable product, e.g. the **Hoodie** imported from WC's sample products
1. Switch to the **Organization** tab and find the **Google Listings & Ads** section.
1. Within the **Google Listings & Ads** section, there should be two fields. Operate the selection field and change its selected option.
1. Click on the **Update** button at the top-right of the page.
1. Refresh the page or switch back to the classical product editor to see if the changed option is saved.
1. Back to edit the same variable product with Product Block Editor.
1. Switch to the **Variations** tab and find the **Variations** section.
1. Click on one of the Edit links within the **Variations** section to go to edit a variation product.
1. With the **General** tab, find the **Google Listings & Ads** section.
1. Repeat the similar test to verify the fields for this variation product.

#### 📌 Test the related filters

1. Go to edit a simple product.
1. Filters the options for each attribute - `"woocommerce_gla_product_attribute_value_options_{$attribute_id}"`
   1. Confirm the **Condition** field has four options: Default, New, Refurbished, and Used.
   1. Add a filter to the **google-listings-and-ads.php** file:
      ```php
      add_filter(
      	'woocommerce_gla_product_attribute_value_options_condition',
      	function ( array $value_options ) {
      		$value_options[ 'broken' ] = 'Broken';
      		return $value_options;
      	}
      );
      ```
   1. Refresh the page.
   1. Check if the **Condition** has the fifth option: Broken.
      ![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/9f2607d6-2492-4f29-a9c2-0c8b2636581e)

### Additional details:

With WooCommerce 8.3.0, the tooltip is not aligned well, and the generic product blocks don't support the `tooltip` attribute.

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/95f90362-66a5-4ffa-b3ff-494e0e4a19bb)

These issues will be resolved in WooCommerce 8.4.0, so this PR won't take care of them.

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/6fc34b81-2f85-4ea4-bd87-cdd47c5b5863)

### Changelog entry
